src/main/java/com/bazel_diff/BazelSourceFileTarget.java

```diff
@@ -18,23 +18,45 @@
     private String name;
     private byte[] digest;
 
+    private void digestLargeFile(MessageDigest finalDigest, FileChannel inChannel) throws IOException {
+        int bufferSize = 10240; // 10kb
+        ByteBuffer buffer = ByteBuffer.allocate(bufferSize);
+        while (inChannel.read(buffer) != -1) {
+            buffer.flip();
+            finalDigest.update(buffer);
+            buffer.clear();
+        }
+    }
+
+    private void digestSmallFile(MessageDigest finalDigest, FileChannel inChannel) throws IOException {
+        long fileSize = inChannel.size();
+        ByteBuffer buffer = ByteBuffer.allocate((int) fileSize);
+        inChannel.read(buffer);
+        buffer.flip();
+        finalDigest.update(buffer);
+    }
+
     BazelSourceFileTargetImpl(String name, byte[] digest, Path workingDirectory)
         throws IOException, NoSuchAlgorithmException {
         this.name = name;
         MessageDigest finalDigest = MessageDigest.getInstance("SHA-256");
         if (workingDirectory != null && name.startsWith("//")) {
             String filenameSubstring = name.substring(2);
             String filenamePath = filenameSubstring.replaceFirst(":", "/");
-            File sourceFile = new File(workingDirectory.toString(), filenamePath);
-            if (sourceFile.isFile() && sourceFile.canRead()) {
-                byte[] buffer = new byte[16384];
-                FileInputStream in = new FileInputStream(sourceFile);
-                int rc = in.read(buffer);
-                while (rc != -1) {
-                    finalDigest.update(buffer, 0, rc);
-                    rc = in.read(buffer);
+            Path absoluteFilePath = Paths.get(workingDirectory.toString(), filenamePath);
+            try (RandomAccessFile sourceFile = new RandomAccessFile(absoluteFilePath.toString(), "r")) {
+                FileChannel inChannel = sourceFile.getChannel();
+                if (inChannel.size() > 1048576) { // 1mb
+                    digestLargeFile(finalDigest, inChannel);
+                } else {
+                    digestSmallFile(finalDigest, inChannel);
                 }
-            }
+                inChannel.close();
+                sourceFile.close();
+                finalDigest.update(digest);
+            } catch (FileNotFoundException e) {}
```

Comment:
Should I log not found files when `verbose` is true?

Reply:
Yes please do